src/web-extensions.c

```diff
@@ -0,0 +1,696 @@
+    {
+      g_dbus_method_invocation_return_gerror (invocation, error);
+      session_close (session, FALSE);
+      return TRUE;
+    }
+  session_register (session);
+
+  xdp_dbus_web_extensions_complete_create_session (object, invocation, session->id);
+
+  return TRUE;
+}
+
+static void
+on_server_exited (GPid pid,
+                  gint status,
+                  gpointer user_data)
+{
+  Session *session = user_data;
+  WebExtensionsSession *we_session = (WebExtensionsSession *)session;
+
+  SESSION_AUTOLOCK (session);
+  we_session->child_pid = -1;
+  we_session->child_watch_id = 0;
+  session_close (session, TRUE);
+}
+
+static gboolean
+array_contains (JsonArray *array,
+                const char *value)
+{
+  guint length, i;
+
+  if (array == NULL)
+    return FALSE;
+
+  length = json_array_get_length (array);
+  for (i = 0; i < length; i++)
+    {
+      const char *element = json_array_get_string_element (array, i);
+      if (g_strcmp0 (element, value) == 0)
+        return TRUE;
+    }
+  return FALSE;
+}
+
+static gboolean
+is_valid_name (const char *name)
+{
+  return g_regex_match_simple ("^\\w+(\\.\\w+)*$", name, 0, 0);
+}
+
+static GStrv
+get_manifest_search_path (void)
+{
+  const char *hosts_path_str;
+  g_autoptr(GPtrArray) search_path = NULL;
+
+  hosts_path_str = g_getenv ("XDG_DESKTOP_PORTAL_WEB_EXTENSIONS_PATH");
+  if (hosts_path_str != NULL)
+    return g_strsplit (hosts_path_str, ":", -1);
+
+  /* By default, use the native messaging search paths of Firefox,
+   * Chrome, and Chromium, as documented here:
+   *
+   * https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Native_manifests#manifest_location
+   * https://developer.chrome.com/docs/apps/nativeMessaging/#native-messaging-host-location
+   */
+
+  search_path = g_ptr_array_new_with_free_func (g_free);
+  /* Add per-user directories */
+  g_ptr_array_add (search_path, g_build_filename (g_get_home_dir (), ".mozilla", "native-messaging-hosts", NULL));
+  g_ptr_array_add (search_path, g_build_filename (g_get_user_config_dir (), "google-chrome", "NativeMessagingHosts", NULL));
+  g_ptr_array_add (search_path, g_build_filename (g_get_user_config_dir (), "chromium", "NativeMessagingHosts", NULL));
+
+  /* Add system wide directories */
+  g_ptr_array_add (search_path, g_strdup ("/usr/lib/mozilla/native-messaging-hosts"));
+  g_ptr_array_add (search_path, g_strdup ("/usr/lib64/mozilla/native-messaging-hosts"));
+  g_ptr_array_add (search_path, g_strdup ("/etc/opt/chrome/native-messaging-hosts"));
+  g_ptr_array_add (search_path, g_strdup ("/etc/chromium/native-messaging-hosts"));
+
+  /* And the same for xdg-desktop-portal's configured prefix */
+  g_ptr_array_add (search_path, g_strdup (LIBDIR "mozilla/native-messaging-hosts"));
+  g_ptr_array_add (search_path, g_strdup (SYSCONFDIR "opt/chrome/native-messaging-hosts"));
+  g_ptr_array_add (search_path, g_strdup (SYSCONFDIR "chromium/native-messaging-hosts"));
```

Comment:
Hm, I think we're going to need to (also?) append the paths of `g_get_system_data_dirs() / {chrome,firefox} / native-messaging-hosts` here, which would include `/var/lib/flatpak/exports/share` (I know Flatpak doesn't currently export these files, but it must, in time for a xdg-desktop-portal release containing this portal)

Reply:
As mentioned in https://github.com/flatpak/xdg-desktop-portal/pull/705#issuecomment-1313572061, I think this addition was a bit dubious, so I'd be inclined to remove them entirely. Both Firefox and Chromium are using hard coded paths.

It's always going to be `/usr/lib{,64}/mozilla` for Firefox, and `/etc/chromium` for Chromium. No matter what install location is used. I think it'd be better to leave them out until we've got an actual plan for confined native messaging hosts.

Reply:
I don't think it makes sense to land a portal that doesn't work with Flatpaks, given that portals are Flatpak infrastructure, so we should figure this before merging.

Reply:
The portal API will work fine for Flatpak packaged browsers (the same as it does for snap packaged browsers). I thought we'd agreed to land support for accessing the host system native messaging servers first, and look at sandboxed native messaging servers as a future improvement.

In particular, you'd need to decide whether you want to support browsers launching native messaging servers through x-d-p (which at present is just unreleased versions of Firefox running in a sandbox), or do you want to support existing browsers too?

No combination of environment variables will get those existing browsers to look for manifests under `/var/lib/flatpak/exports/share`, so you'd need to install to one of the locations they currently use to be found. And if you do that, then there is no need for the extra search paths in x-d-p.

Reply:
The most immediate question raised by adding support for exporting from Flatpaks is that the current JSON metadata format only allows specifying a binary path, not with arguments. So the approach taken in https://github.com/flatpak/flatpak/blob/main/common/flatpak-dir.c#L7589 to build a `flatpak run --command` and substitute into the Exec= line of a .desktop file won't work. Flatpak does export binary wrappers but only one per app, which is presumed to be the default command to launch the main app. Imagining, eg, KeePassXC, you wouldn't want the default binary when you click Run in the app store to be the native messaging host - you'd want the UI to launch.

So we have a question to answer - do we need to work out a way to have Flatpak export "auxilliary" binaries (eg `app.id-cmdname`) which I think would be useful in some other circumstances, so as not to mess around with the arguments which are sent to the messaging host, a special-case binary wrapper which is exported alongside each host .json, or do we want to add syntax to the file (eg an exec cmdline rather than just a binary path) so that the portal knows how to exec it via Flatpak? Requiring a format/semantic change means that non-sandboxed browsers will be unable to work with sandboxed apps/hosts, although it may already be the case that unless Flatpak puts them into the system paths for such things (which I'd say we're... unlikely to do?) they won't be found anyway.